File: tools/linear-sync/src/redact.ts
// redact.ts is this workspace's canary-scan and safe-uncertain-outcome
// contract (Plan 01-27; CONTEXT D-19/D-20; T-01-41) -- the TypeScript-side
// sibling of internal/security/redact.go. adapter.ts's createOperation/
// updateOperation route every raw create/update failure (a network
// timeout, an aborted request, a partial GraphQL error mid-mutation, or any
// other exception the official @linear/sdk or this adapter's own mandatory
// readback throws) through safeError before it can ever reach a returned
// MutationOutcome (protocol.ts), cli.ts's NDJSON output line, or process
// stderr. safeError never reads a raw exception's message, stack, cause
// content, or any attached property (a client instance, environment map,
// request body, or header set): only its constructor-assigned `.name` -- a
// small, fixed, safe classification the JS runtime or SDK itself assigns,
// never attacker- or credential-influenced free text -- ever informs the
// returned diagnostic's `code`, which is always one of exactly three fixed
// string constants. scanCanary/scanCanaryAll are this module's own offline
// proof mechanism, byte-for-byte matching internal/security/redact.go's
// CanaryToken and forbidden-pattern list, so redact.test.ts (and
// mutation.test.ts) can plant a unique fake secret inside a hostile raw
// error and prove it never survives safeError's normalization.

import type { RequestContext } from "./errors.js";
import type { TransportDiagnostic } from "./protocol.js";

/**
 * CANARY_TOKEN mirrors internal/security/redact.go's CanaryToken exactly
 * (byte-for-byte identical value) -- a unique, obviously-fake secret-shaped
 * byte sequence this module's own test suite plants into a controlled raw
 * error to prove scanCanary actually detects leaked secret material, never
 * only observing already-clean output. It must never appear in any real
 * committed or generated repository artifact.
 */
export const CANARY_TOKEN = "GOLC_FAKE_SECRET_CANARY_4f9c2e6b1a7d3f809c21";

/**
 * FORBIDDEN_PATTERNS mirrors internal/security/redact.go's
 * forbiddenPatterns exactly: the exact CANARY_TOKEN plus the fixed set of
 * common secret-shaped substrings (a raw Linear credential declaration, a
 * bearer auth header, and API-key prefixes) that must never appear in
 * emitted diagnostics or committed/generated artifacts, independent of
 * whether CANARY_TOKEN itself was ever planted.
 */
const FORBIDDEN_PATTERNS: readonly string[] = [CANARY_TOKEN, "LINEAR_API_KEY=", "Bearer ", "sk-", "lin_api_"];

/**
 * scanCanary reports the first forbidden token found in data, or undefined
 * when data is clean -- the exact TypeScript-side mirror of
 * internal/security/redact.go's ScanCanary. It operates on the full text of
 * whatever surface a caller passes (a rendered diagnostic, an NDJSON output
 * line, a committed fixture's own JSON), never a special-cased subset.
 */
export function scanCanary(data: string): string | undefined {
  for (const token of FORBIDDEN_PATTERNS) {
    if (data.includes(token)) {
      return token;
    }
  }
  return undefined;
}

/**
 * CanaryViolation names one output surface that leaked a forbidden token,
 * mirroring internal/security/redact.go's CanaryViolation.
 */
export interface CanaryViolation {
  source: string;
  token: string;
}

/**
 * scanCanaryAll scans every named source in sources and returns every
 * violation, sorted by source name for deterministic reporting -- the exact
 * TypeScript-side mirror of internal/security/redact.go's ScanCanaryAll. An
 * empty array means every source is clean.
 */
export function scanCanaryAll(sources: Readonly<Record<string, string>>): CanaryViolation[] {
  const names = Object.keys(sources).sort();
  const violations: CanaryViolation[] = [];
  for (const name of names) {
    const token = scanCanary(sources[name] ?? "");
    if (token !== undefined) {
      violations.push({ source: name, token });
    }
  }
  return violations;
}

// ---------------------------------------------------------------------------
// safeError -- the sole producer of an "unknown" MutationOutcome's
// diagnostic (adapter.ts, Plan 01-27).
// ---------------------------------------------------------------------------

/**
 * MUTATION_FAILURE_CODES is the exhaustive, fixed set of safe diagnostic
 * codes safeError may ever assign -- never a raw exception's own message or
 * any other free-text content (CONTEXT D-20's "secret values must never
 * appear in previews, logs, errors" extended here to every mutation
 * failure, not only credentials).
 */
const MUTATION_FAILURE_CODES = {
  timeout: "LINEAR_MUTATION_TIMEOUT",
  network: "LINEAR_MUTATION_NETWORK_ERROR",
  unknown: "LINEAR_MUTATION_UNCERTAIN",
} as const;

/**
 * hasStructuredCause reports whether error carries an object-shaped
 * `.cause` property (the shape Node's own fetch-style network failures
 * use) -- inspected only for its type, never its content, so this
 * classification step itself can never read or forward a raw value.
 */
function hasStructuredCause(error: Error): boolean {
  const cause = (error as { cause?: unknown }).cause;
  return typeof cause === "object" && cause !== null;
}

/**
 * classifyMutationFailure maps a raw, untrusted create/update failure to
 * one of MUTATION_FAILURE_CODES's three safe codes. It inspects only an
 * Error's constructor-assigned `.name` ("AbortError"/"TimeoutError" for a
 * cancelled/timed-out request, "TypeError" with a structured `.cause` for a
 * lower-level network failure) and never reads `.message`, `.stack`, or any
 * other property -- a hostile or malformed raw error can never smuggle
 * arbitrary content into the returned code by construction, since the
 * return value is always one of exactly three fixed string constants.
 */
function classifyMutationFailure(error: unknown): string {
  if (error instanceof Error) {
    if (error.name === "AbortError" || error.name === "TimeoutError") {
      return MUTATION_FAILURE_CODES.timeout;
    }
    if (error.name === "TypeError" && hasStructuredCause(error)) {
      return MUTATION_FAILURE_CODES.network;
    }
  }
  return MUTATION_FAILURE_CODES.unknown;
}

/**
 * safeError converts one raw, untrusted create/update failure (any
 * exception thrown by the official @linear/sdk's mutation call, or by this
 * adapter's own mandatory readback immediately following it) into the
 * exact allowlisted TransportDiagnostic surface protocol.ts declares:
 * operation/endpoint come only from the caller's own RequestContext (never
 * the error itself), and code is always one of classifyMutationFailure's
 * three fixed constants. No raw exception message, stack trace, GraphQL
 * query/variables, header value, credential, client instance, environment
 * map, or request body ever reaches the returned diagnostic -- this is the
 * sole producer of an "unknown" MutationOutcome's diagnostic (adapter.ts).
 */
export function safeError(error: unknown, context: RequestContext): TransportDiagnostic {
  const diagnostic: TransportDiagnostic = { operation: context.operation, code: classifyMutationFailure(error) };
  if (context.endpoint !== undefined) {
    diagnostic.endpoint = context.endpoint;
  }
  return diagnostic;
}

File: internal/security/redact.go
// Package security is the single centralized owner of GOLC's secret-safe
// diagnostic and canary-scan contract (CONTEXT D-19/D-20, T-01-18): every
// caller that needs to render a sensitive declaration's presence, redact a
// value before it can be logged, or prove that no fake-secret byte leaked
// into stdout, stderr, a generated schema, a committed map, a report, a
// manifest, or a ZIP archive goes through this package rather than
// re-implementing its own ad hoc masking.
//
// SafeDiagnostic is deliberately narrow: a stable machine-readable code, a
// safe message, and an optional map of already-safe named fields. No caller
// may attach a raw environment map, HTTP header set, config struct, or
// error/exception value to it -- there is no field shape that would even
// accept one. Every field value SafeDiagnostic.String renders is re-passed
// through Redact, so a caller cannot bypass the canary/pattern scan by
// constructing a SafeDiagnostic directly with an already-leaked value.
package security

import (
	"fmt"
	"sort"
	"strings"
)

// CanaryToken is a unique, obviously-fake secret-shaped byte sequence this
// package's own test suite plants into a controlled buffer to prove
// ScanCanary actually detects leaked secret material, rather than only ever
// observing already-clean output. It must never appear in any real
// committed or generated repository artifact.
const CanaryToken = "GOLC_FAKE_SECRET_CANARY_4f9c2e6b1a7d3f809c21"

// forbiddenPatterns are the byte-level tokens ScanCanary rejects: the exact
// CanaryToken plus a fixed set of common secret-shaped substrings (Linear
// credential declarations, bearer auth headers, and API-key prefixes) that
// must never appear in emitted diagnostics or committed/generated
// artifacts, independent of whether the CanaryToken itself was ever
// planted.
var forbiddenPatterns = []string{
	CanaryToken,
	"LINEAR_API_KEY=",
	"Bearer ",
	"sk-",
	"lin_api_",
}

// ScanCanary reports the first forbidden token found in data, or "" when
// data is clean. It operates on raw bytes so stdout/stderr captures,
// generated schema/map/report/manifest files, and raw ZIP archive bytes are
// all scanned identically -- no output surface receives special-cased
// treatment.
func ScanCanary(data []byte) string {
	text := string(data)
	for _, token := range forbiddenPatterns {
		if strings.Contains(text, token) {
			return token
		}
	}
	return ""
}

// CanaryViolation names one output surface that leaked a forbidden token.
type CanaryViolation struct {
	Source string
	Token  string
}

// ScanCanaryAll scans every named source in sources and returns every
// violation, sorted by source name for deterministic reporting. A nil or
// empty return value means every source is clean. Callers use this instead
// of repeated ScanCanary calls so a failure can name exactly which output
// surface leaked.
func ScanCanaryAll(sources map[string][]byte) []CanaryViolation {
	names := make([]string, 0, len(sources))
	for name := range sources {
		names = append(names, name)
	}
	sort.Strings(names)

	var violations []CanaryViolation
	for _, name := range names {
		if token := ScanCanary(sources[name]); token != "" {
			violations = append(violations, CanaryViolation{Source: name, Token: token})
		}
	}
	return violations
}

// SetState renders whether a sensitive declaration (an environment variable
// such as LINEAR_API_KEY) is populated without ever exposing its value
// (CONTEXT D-19/D-20): "<set>" when a non-blank value is present, "<unset>"
// otherwise. This is the only rendering form provenance/status output may
// use for a declared-secret field.
func SetState(value string) string {
	if strings.TrimSpace(value) == "" {
		return "<unset>"
	}
	return "<set>"
}

// redactedMarker replaces any value ScanCanary flags. Redact never returns
// even a truncated fragment of the original bytes for a flagged value, so a
// coding mistake elsewhere can never leak a real secret through a
// "redacted" value that still contains part of it.
const redactedMarker = "<redacted>"

// Redact returns value unchanged when it is already safe (ScanCanary finds
// no forbidden token in it); otherwise it returns the fixed redactedMarker.
func Redact(value string) string {
	if ScanCanary([]byte(value)) != "" {
		return redactedMarker
	}
	return value
}

// SafeDiagnostic is the only allowlisted diagnostic shape this repository's
// CLI, configuration, and adapter code may build for anything
// environment/header/config/exception shaped: a stable code, a safe
// message, and an optional map of safe named fields. Every field value is
// re-redacted at render time (String), so constructing a SafeDiagnostic
// directly with an unvetted value can never leak it.
type SafeDiagnostic struct {
	Code    string
	Message string
	Fields  map[string]string
}

// String renders d as one stable, redaction-safe line:
// "<code>: <message> (field=value, field=value)", with fields sorted by
// name for deterministic output. The parenthesized field list is omitted
// entirely when d.Fields is empty.
func (d SafeDiagnostic) String() string {
	var builder strings.Builder
	builder.WriteString(d.Code)
	builder.WriteString(": ")
	builder.WriteString(d.Message)
	if len(d.Fields) == 0 {
		return builder.String()
	}

	names := make([]string, 0, len(d.Fields))
	for name := range d.Fields {
		names = append(names, name)
	}
	sort.Strings(names)

	parts := make([]string, 0, len(names))
	for _, name := range names {
		parts = append(parts, fmt.Sprintf("%s=%s", name, Redact(d.Fields[name])))
	}
	builder.WriteString(" (")
	builder.WriteString(strings.Join(parts, ", "))
	builder.WriteString(")")
	return builder.String()
}
